### ISSUE TITLE
"Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock'"

### ISSUE THREAD

**Issue by danielbachhuber** [travis-ci/travis-ci#6842]
My builds are failing on this error:

```
mysql -e 'CREATE DATABASE wp_cli_test;' -uroot
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)
```

See https://travis-ci.org/wp-cli/wp-cli/jobs/174318358 for a representative job.

This issue is specific to Travis CI and new within the last ~8 hours. Happy to help debug further if you'd like to point me in a direction.

**Comment by photodude** [comment]
Also seeing this on the [Container-based Ubuntu Trusty Public Beta](https://blog.travis-ci.com/2016-11-08-trusty-container-public-beta/) 
```php
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)
```
 Reference #5695

**Comment by danielbachhuber** [comment]
Thanks @photodude. Didn't realize infrastructure changes rolled out today.

**Comment by photodude** [comment]
@danielbachhuber Looks like your running Trusty beta. drop `dist: trusty` from your travis.yml to fix the issue. @szepeviktor made that change.

**Comment by danielbachhuber** [comment]
@photodude Right. Can we expect the issue to be fixed any time soon, or should I revert in all of the relevant places?

**Comment by photodude** [comment]
I would revert for now for live repos needing to do continuous testing since Trusty is beta. Test your repo with trusty beta in a separate repo and plan to migrate some time after Q1-17 when it's officially released.

Trusty likely doesn't benefit you unless you need something like HHVM 3.9+ (current hhvm is 3.15)  which is trusty specific.

**Comment by danielbachhuber** [comment]
👍  Thanks for your help, @photodude

**Comment by photodude** [comment]
You are welcome @danielbachhuber, Glad your up and testing again. Hopefully the trusty issues can be sorted out soon (I'm hoping it will be faster for testing HHVM)

PS Please keep this issue open until it's resolved on Container-based Trusty Beta.

**Comment by lesterchan** [comment]
I symlinked the socket and it works.
```
sudo mkdir -p /var/run/mysqld
sudo ln -s /run/mysql-5.6/mysqld.sock /var/run/mysqld/mysqld.sock
```

**Comment by photodude** [comment]
@lesterchan symlinking the socket like that doesn't work on the Container-based Trusty Beta as sudo is disabled on the Containers.

**Comment by lesterchan** [comment]
@photodude ops, yeap forgot to add that. My .travis.yml has sudo required.

**Comment by szepeviktor** [comment]
@danielbachhuber We you shim it for the time being:

```bash
alias mysql='mysql --socket=/run/mysql-5.6/mysqld.sock'
```

**Comment by photodude** [comment]
@lesterchan I have no issue with the socket on sudo required trusty. For me this is an issue only with the Container-based Trusty Beta.

**Comment by arogachev** [comment]
I'm facing the same issue. Fixed it by installing and using [MySQL 5.6](https://docs.travis-ci.com/user/database-setup/#MySQL-5.6):

```yaml
addons:
  apt:
    packages:
      - mysql-server-5.6
      - mysql-client-core-5.6
      - mysql-client-5.6
```

But along with it we must change user to `root` because `travis` is not supported. Also default is MySQL 5.5 in both `precise` and `trusty`. We don't want to bump the version just to fix it. So I'm not happy with this solution, looking for better one. Will try to use `alias`.

**Comment by arogachev** [comment]
This:

```
alias mysql='mysql --socket=/run/mysql-5.5/mysqld.sock'
```

didn't work for me. Gives the following error:

```
ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)
```

I checked contents of both `/run` and `/var/run` folders with `ls -la`, there are no mysql related files or folders at all.

**Comment by szepeviktor** [comment]
Please see our successful builds  https://travis-ci.org/wp-cli/wp-cli/builds/174754228


-- 
+36204242498
Ezen a készüléken sok az elütés. 
Elnézést!

**Comment by arogachev** [comment]
To continue using MySQL 5.5 and prevent upgrading to 5.6 we can install missing mysql-server-5.5 package:

```yaml
addons:
  apt:
    packages:
      - mysql-server-5.5
```

Now the error is gone. As a temporary fix this is OK, but I hope it will be fixed in a better way, for example by having this package installed by default.

**Comment by photodude** [comment]
@arogachev I installed mysql 5.6 since that is how I have sudo required trusty set up and working. Already [using mysql root](https://github.com/photodude/joomla-cms/blob/patch-10/build/travis/unit-tests.sh) as that is required for the sudo required trusty, I only changed the lines to switch to the Container-based Trusty Beta.

This is a Container-based Trusty Beta specific issue

@szepeviktor 's suggestion of aliasing the socket does get past this issue, but reveals the issue noted in #6853

**Comment by photodude** [comment]
@szepeviktor for your workaround on [wp-cli](https://travis-ci.org/wp-cli/wp-cli/builds/174754228) you have to also specify `dist: precise` under the php 5.3 as Trusty only default supports php 5.4+

Additionally, I [noticed with the alias workaround](https://travis-ci.org/photodude/joomla-cms/jobs/175096904#L1094-L1095) there are PDO issues
```php
PDOException: SQLSTATE[HY000] [2002] No such file or directory
```

**Comment by arogachev** [comment]
@photodude Maybe there are some nuances on Container-based Ubuntu Trusty, but I get the same error on Standard Ubuntu Trusty using config:

```yaml
dist: trusty
sudo: required
```

Without additional installation I mentioned above the error persists.

**Comment by photodude** [comment]
@arogachev we do not have this issue on sudo required trusty [for Joomla](https://travis-ci.org/joomla/joomla-cms/jobs/175083433) or [with doctrine dbal](https://travis-ci.org/doctrine/dbal/jobs/173903364) (note only hhvm is tested via sudo required)  I'm only seeing this issue with Container-based Trusty Beta. 

Is this a new issue you are seeing on sudo required trusty or something from a while back?

**Comment by photodude** [comment]
@arogachev [I ran a new test](https://travis-ci.org/photodude/joomla-cms/jobs/175110744), I can confirm what you are seeing in sudo required trusty. Seems to be in issue if the group is not set to edge
add this line to the top of your travis.yml and the issue should go away as it did for [my edge test](https://travis-ci.org/photodude/joomla-cms/jobs/175113595)
```php
group: edge
```
Although I'm still getting a related `PDOException issue` on php with `group: edge`
```PDOException: SQLSTATE[HY000] [2002] Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)```

**Comment by arogachev** [comment]
@photodude It happens now, first time I saw this error was yesterday. You can see according build job and config [here]((https://travis-ci.org/yiisoft/yii2/jobs/174692067).

**Comment by szepeviktor** [comment]
> @szepeviktor 's suggestion of aliasing the socket does work arround this issue, but reveals the issue noted in #6853

@photodude Of course you have to also tell PHP PDO to connect on another socket!

**Comment by szepeviktor** [comment]
Maybe here http://php.net/manual/en/ref.pdo-mysql.php#ini.pdo-mysql.default-socket

**Comment by photodude** [comment]
@szepeviktor the socket issue on trusty with mysql(i) PDO can be fixed by changing the DSN host from `localhost` to `127.0.0.1`

but results in
```php
PDOException: SQLSTATE[28000] [1045] Access denied for user 'root'@'localhost' (using password: NO)
```

**Comment by szepeviktor** [comment]
You should connect through the socket file not the TCP socket. 



-- 
+36204242498
Ezen a készüléken sok az elütés. 
Elnézést!

**Comment by szepeviktor** [comment]
I think it is not an uds/tcp problem.

Joomla expects a passwordless root account on Mysql but this new TravisCI container may not have one.

Excuse me. 


-- 
+36204242498
Ezen a készüléken sok az elütés. 
Elnézést!

**Comment by photodude** [comment]
The following workaround seems to work for fixing the the related PDO issue
```php
before_install:
  - "sed -i '/^[a-z_]*mysql.*socket/s:$: /run/mysql-5.6/mysqld.sock:' ~/.phpenv/versions/5.6*/etc/php.ini"
```

**Comment by jakub300** [comment]
It looks like there is default password `travis` for root. You can reset it by running `mysql -e 'SET PASSWORD FOR "root"@"localhost" = PASSWORD("");' -uroot` in `before_install`.

**Comment by photodude** [comment]
Thanks @jakub300, that might help has a work around. 

Travis has confirmed the socket issues and will be fixing them after the holidays.

**Comment by geerlingguy** [comment]
I'm hitting this starting today on the Trusty beta container build on travis-ci.com (for a private repo)... and when I debugged the build (SSH in), and ran the exact same MySQL command that was failing, it worked fine.

Then I kept restarting the same build, and fourth time was a charm. It seems to hit at random; like the socket is fine much of the time, but sometimes it or MySQL just goes away for a bit.

**Comment by cebe** [comment]
This is still happening, seems the mysql server fails to start for some reason. There is no detailed output, just this:

> $ sudo service mysql start
> start: Job failed to start
> https://travis-ci.org/yiisoft/yii2/jobs/198069379#L111

**Comment by photodude** [comment]
@cebe I think that's a separate issue from the trusty beta socket issue here

**Comment by cebe** [comment]
@photodude why do you think it is different? At least the error messages on commands are the same: https://travis-ci.org/yiisoft/yii2/jobs/198069379#L495

**Comment by photodude** [comment]
@cebe, awe I see, I just looked at the info you posted on GitHub and not at the Travis build. I also didn't expect you to have changed to the trusty beta after the work we did on Yii where I listed that it would be after 1Q17 when the switch to the trusty container would be advisable.

**Comment by rbs7816** [comment]
I have a solution to this problem and it works on my WSL 
Type these:

1. sudo apt-get update
2. sudo sed -i 's$<listen>.*</listen>$<listen>tcp:host=localhost,port=0</listen>$' /etc/dbus-1/session.conf
3. sudo service mysql start....

Your MySQL service is enabled....

**Comment by danepowell** [comment]
I might have a solution courtesy of Travis support. The problem might be that MySQL isn't fully initialized, so add a sleep statement before you call it for the first time...

```yaml
before_install:
  # Allow MySQL to finish starting
  - sleep 3
```

**Comment by cebe** [comment]
On this build I have package installation and other things before accessing mysql that take over 20 seconds: https://travis-ci.org/yiisoft/yii2/jobs/199528768 so that does not really sound useful.

**Comment by alecha** [comment]
Same here, the sleep workaround doesn't work

**Comment by danepowell** [comment]
Yeah sorry folks, I confirmed that the sleep workaround doesn't work and will get back in touch with Travis support today. It's just hard to verify one way or the other since this problem occurs so randomly.

Specifically, I replaced the sleep statement with this to verify that no matter how long you wait, the MySQL server never comes up:

```bash
while ! mysqladmin ping --silent; do sleep 5; echo "Waiting for MySQL..."; done
```

**Comment by photodude** [comment]
@alecha This is a trusty beta container issue. `sudo: required` trusty was fixed several months ago. 
you must have `sudo:false` and `dist: trusty` to run into this issue.

**Comment by photodude** [comment]
@danepowell I've been in touch with Travis support since the beginning of this beta container issue. There is something slightly different in the beta container from what solved the socket problem on sudo: required trusty. Still waiting for a final solution to be implemented for the beta container trusty.

**Comment by danepowell** [comment]
Yeah adding the apt package lines for mysql-5.6 does nothing if you're already on a containerized build. The reason that worked for you @alecha is that `sudo: required` switched you to the non-container build.

**Comment by alecha** [comment]
All of you guys are right, I actually had to switch back to `sudo:required` otherwise we were blocked with the build

**Comment by geerlingguy** [comment]
See related: https://github.com/travis-ci/travis-ci/issues/6461

**Comment by Krinkle** [comment]
Related:
* https://github.com/travis-ci/travis-ci/issues/6982
* https://github.com/travis-ci/travis-ci/issues/6853
* https://github.com/travis-ci/travis-ci/issues/5837

**Comment by maartenwodan** [comment]
**Solution for PHP**:
 - Create a `travisphp.ini` in your root *(I've had issues with the filename `.travisphp.ini` before, and also file names with dashes like `travis-php.ini` do not seem to work)*.
- Fill it with the following, when using `mysql-5.6`:

```ini
mysql.default_socket = "/run/mysql-5.6/mysqld.sock"
mysqli.default_socket = "/run/mysql-5.6/mysqld.sock"
pdo_mysql.default_socket = "/run/mysql-5.6/mysqld.sock"
```
- Add the following to the `.travis.yml`:

```yml
before_script:
  - phpenv config-add travisphp.ini
```

**Comment by danepowell** [comment]
Travis support indicates that this should no longer be an issue with the latest Trusty builds... we'll test and see.

**Comment by kiklop74** [comment]
Started 3 builds  and one of them failed like this. See for yourself:

https://travis-ci.org/kiklop74/moodle-local_xray/jobs/255790946

**Comment by shioyama** [comment]
Also seeing this having been switched over to Trusty. I don't really understand Travis bumped everyone to Trusty without resolving this issue first, since I have to imagine it will affect so many users.

I've just switched back to Precise with `dist: precise` for now.

**Comment by benpbolton** [comment]
Just fought a bit with this. Here was our solution:

Ensure:
```
services:
  - mysql
``` 
... is present in your `.travis.yml` since [the docs](https://docs.travis-ci.com/user/database-setup/#Starting-Services) says

> Travis CI environments do not start services by default, to make more RAM available to build scripts.

... annnnd... if you want to ensure that the service had enough time to spin up before you start interacting with it, you might give something like this a go:

```
    # Wait for mysql service to respond
    - which mysql && until mysql -u root -e "show status" &>/dev/null; do sleep 1; done
```

**Comment by Krinkle** [comment]
For wikimedia/mediawiki, containers (`sudo: false`) with MySQL (`services: - mysql`) now work fine as-is and without additional workarounds or timers.

**Comment by shioyama** [comment]
Suddenly seeing these errors again... anybody else having issues with this?

**Comment by shioyama** [comment]
Seems to have fixed itself, not seeing any more failures.

**Comment by shioyama** [comment]
Never mind, still happening. Is no-one else having this issue? Really annoying to have to restart builds all the time.

**Comment by benpbolton** [comment]
hey @shioyama , we definitely haven't run into this in the last year or so :( Perhaps suspect odd behavior for postgres vs mysql env? Eg. perhaps the mysql service doesn't launch when  `DB=postgres` or something similar? So the 
```
    - which mysql && until mysql -u root -e "show status" &>/dev/null; do sleep 1; done
```
is just looping forever in that context. You could test if $DB == mysql before waiting for it if that was helpful

**Comment by adteulade** [comment]
Hey we were having the same issue since yesterday. After some investigations it appears that the ubuntu trusty isn't somehow the default OS used for build. Using :
```
dist: trusty
```
in your .travis.yml fixed this issue (after ```language: php```). 
The weirdest thing is that when restarting master build, tests were working but if we were creating a new branch from master without changing anything we had the issue

**Comment by BanzaiMan** [comment]
https://travis-ci.community/t/prs-started-failing-on-6-12/3707/2

We are moving to Xenial as the default, and we do not start DB by default.

https://blog.travis-ci.com/2019-04-15-xenial-default-build-environment#1-services-support

You need to explicitly instruct that it is needed.

```yaml
services:
  - mysql
```
